app/src/main/java/org/jellyfin/androidtv/ui/livetv/LiveTvGuideActivity.java

```diff
@@ -767,26 +797,36 @@ protected void onPostExecute(Void aVoid) {
 
     private int currentCellId = 0;
 
+    private GuideChannelHeader getChannelHeader(Context context, ChannelInfoDto channel, boolean focusable){
+        return new GuideChannelHeader(context, null, channel, focusable);
+    }
+
     private LinearLayout getProgramRow(List<BaseItemDto> programs, String channelId) {
 
         LinearLayout programRow = new LinearLayout(this);
 
         if (programs.size() == 0) {
             if (mFilters.any()) return null; // don't show rows with no program data
 
-            BaseItemDto empty = new BaseItemDto();
-            int duration = ((Long)((mCurrentLocalGuideEnd - mCurrentLocalGuideStart) / 60000)).intValue();
-            empty.setName("  <No Program Data Available>");
-            empty.setChannelId(channelId);
-            empty.setStartDate(TimeUtils.convertToUtcDate(new Date(mCurrentLocalGuideStart)));
-            empty.setEndDate(TimeUtils.convertToUtcDate(new Date(mCurrentLocalGuideStart+(duration*60000))));
-            ProgramGridCell cell = new ProgramGridCell(this, this, empty);
-            cell.setId(currentCellId++);
-            cell.setLayoutParams(new ViewGroup.LayoutParams(duration * PIXELS_PER_MINUTE, ROW_HEIGHT));
-            cell.setFocusable(true);
-            programRow.addView(cell);
-            cell.setLast();
-            cell.setFirst();
+            int minutes = ((Long)((mCurrentLocalGuideEnd - mCurrentLocalGuideStart) / 60000)).intValue();
+            int slot = 0;
+            do {
+                BaseItemDto empty = new BaseItemDto();
+                empty.setName("  " + getString(R.string.no_program_data));
+                empty.setChannelId(channelId);
+                empty.setStartDate(TimeUtils.convertToUtcDate(new Date(mCurrentLocalGuideStart + ((30*slot) * 60000))));
+                empty.setEndDate(TimeUtils.convertToUtcDate(new Date(mCurrentLocalGuideStart + ((30+(slot+1)) * 60000))));
+                ProgramGridCell cell = new ProgramGridCell(this, this, empty, false);
+                cell.setId(currentCellId++);
+                cell.setLayoutParams(new ViewGroup.LayoutParams(30 * PIXELS_PER_MINUTE, ROW_HEIGHT));
+                cell.setFocusable(true);
+                programRow.addView(cell);
+                if (slot == 0)
+                    cell.setFirst();
+                if (slot == (minutes / 30) - 1)
+                    cell.setLast();
+                slot++;
+            } while((30*slot)*60000 < minutes*60000);
```

Comment:
It shouldn’t be necessary to multiply both sides by 60000 here.

Reply:
Agree. Didn't realize the redundancy. Fixed.